src/Shared/AssemblyNameExtension.cs

```diff
@@ -186,7 +186,23 @@ internal static AssemblyNameExtension GetAssemblyNameEx(string path)
 #if !FEATURE_ASSEMBLYLOADCONTEXT
             try
             {
-                assemblyName = AssemblyName.GetAssemblyName(path);
+                try
+                {
+                    assemblyName = AssemblyName.GetAssemblyName(path);
+                }
+                catch (OutOfMemoryException)
+                {
+                    // Force a full GC and try one more time
+                    // https://github.com/microsoft/msbuild/issues/5150
+                    GC.GetTotalMemory(forceFullCollection: true);
```

Comment:
This is not making any sense. It's not like GC hasn't tried to collect as much as it can before raising the OOM exception. If such a pattern was helpful, it would be the top level statement of every non-trivial program.

Reply:
> It's not like GC hasn't tried to collect as much as it can before raising the OOM exception.

As best I can tell, there is no guarantee that a full GC will run before throwing an OOM. This code is copied from dotnet/roslyn where we have seen similar edge cases that required manual GC:

https://github.com/dotnet/roslyn/blob/2f6273b0e2f219a12166bf09b8e51a50ded050e5/src/Workspaces/Core/Portable/TemporaryStorage/TemporaryStorageServiceFactory.MemoryMappedInfo.cs#L134-L162

Reply:
Wow, that's surprising.

> As best I can tell, there is no guarantee that a full GC will run before throwing an OOM.

@Maoni0, is this so?

Reply:
it's possible, yes, if we determined that doing a full GC would not be productive (ie, not reduce heap size) we don't always do a full GC. this is indicated if you do !ao. 
note that this also means we **have tried** doing a full GC in the recent past and discovered either your live data size keeps not reducing or you had enough pins that we simply couldn't not compact them away.

Reply:
So it would be correct to say a full GC may not have run, but if not, it wouldn't be helpful anyway? I'd be inclined to not run it needlessly.

Reply:
exactly.

Reply:
I've hit this issue a bunch of times this morning, but I haven't hit it during any of the times I opened the Command Window in VS and repeatedly issued the Tools.ForceGC command during the build... If this continues for a few more hours I'm thinking it's not a coincidence.